### DISCUSSION TITLE
TAB Separators: Editing Problems and File Corruption Risks

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#24] [Editors Support]
The _PML Reference Manual_ mentions for [`table_data`][table_data]:

> Cell values in a row are separated by a comma or a TAB character.

This might be a problem with many editors, if the user (or a PML syntax) has set it to use spaces for indentation, because the editor will automatically convert all Tabs to spaces, both during typing as well as at save time.

Probably many users will prefer indenting using spaces, so they won't be able to the TAB separator but only commas, which might be a problem in some cases I guess, e.g. if some cells include commas.

Although I understand that this approach mimics the common separators used in CSV files, you might consider adding an alternative separator to the TAB character, e.g. a pipe `!`.

Also, bear in mind that keeping the TAB could lead to many problems, e.g. a contributor editing a source document that uses TABs might inadvertently get corrupted at save time if the editor replaces al TABs with spaces. Because of how widespread this feature is in modern editors, preserving the TAB as a separator (in `table_data` and any other nodes) is risky, IMO.

[table_data]: https://www.pml-lang.dev/docs/reference_manual/index.html#node_table_data

**Comment by pml-lang** [comment]
> This might be a problem with many editors, if the user (or a PML syntax) has set it to use spaces for indentation

Yes, depending on the editor's configuration. In the case of `table_data`, tabs are not used to indent, just to separate fields. However, some (most?) editors (e.g. VSCode) replace _all_ tabs with spaces (even when they are not used for indentation). If somebody uses `table_data` in such an editor, he/she cannot easily insert a real tab. For example, I use `Alt 009` on my keyboard.

Tabs have the advantage of being pretty much never used in normal text and they "mimic the common separators used in CSV files". But tabs are also indistinguishable from spaces, and difficult to insert if the editor replaces them with spaces.

> a contributor editing a source document that uses TABs might inadvertently get corrupted at save time

Yes, that's a real worst-case-scenario. Also, there is a high risk that the author will even not be aware of the problem.

> you might consider adding an alternative separator to the TAB character, e.g. a pipe

Yes, the comma as alternative is not enough. We could add a few alternatives, but that causes problems if any table cell contains any separator without escaping it.
Therefore IMO we must add an attribute that allows the user to explicitly specify the cell separator (default = comma). It's already on my todo list.

It would also be useful to add an attribute to specify the record separator (default = new line), because that would be useful if cells contain new lines.

Note: All these problems don't exist in the `table` node. But the `table` node (very similar to a HTML table) is more cumbersome to write.

**Reply by tajmone** [reply]
> Yes, the comma as alternative is not enough.

Unless it's possible to enclose a cell within double quotes;

    First cell, "Second cell, quoted", Third

> We could add a few alternatives, but that causes problems if any table cell contains any separator without escaping it.

The vertical bar (aka pipe, vbar, pike) `|` should be a fairly safe choice, due its rare usage in normal text. Other candidates could be the broken `¦` or the double `‖` pipe variants, but their insertion requires using Alt Codes, and the latter might not be so familiar to many users, nor represented by all fonts (especially monospace).

The problem in finding a good separator char is that you want it to be familiar, easy to type, and covered by all fonts, but not commonly used in ordinary text; but characters which are not commonly used are rarely present on most keyboards, and not guaranteed to be covered by all fonts.

> Therefore IMO we must add an attribute that allows the user to explicitly specify the cell separator (default = comma). It's already on my todo list.

This is an excellent solution. Asciidoctor also allows this in [CVS/TVS/DSV tables] — for more details:

- [Asciidoctor Manual: Tables » Data table formats]
- [RFC 4180] — Common Format and MIME Type for Comma-Separated Values (CSV) Files

> It would also be useful to add an attribute to specify the record separator (default = new line), because that would be useful if cells contain new lines.

The `[nl]` node can already be exploited in similar constrained contexts, but it only adds a line break, it doesn't force a new paragraph. Will the record separator also cause just a new line (`<br />`) or will it split paragraphs?

Probably for `[table_data` a newline is more than enough, since for more complex uses the `[table` node should be used.



[table_data]: https://www.pml-lang.dev/docs/reference_manual/index.html#node_table_data

[CVS/TVS/DSV tables]: https://docs.asciidoctor.org/asciidoc/latest/tables/data-format/#delimiter-separated-values "Asciidoctor Manual"

[Asciidoctor Manual: Tables » Data table formats]: https://docs.asciidoctor.org/asciidoc/latest/tables/data-format/#data-table-formats
[RFC 4180]: https://tools.ietf.org/html/rfc4180
